README.md
# configured_mail_sender
A Python module that simplifies sending emails by keeping server and credential information in configuration files.


CHANGELOG.md
# Change log for configured_mail_sender

## 1.0.0
Initial version.
* Release: 26 Sep 2023
             
## 1.0.1
Fix error in finding user creds file.
* Release: 28 Oct 2023

## 1.0.2
Ignore 'name' in user credentials. Specifying an alternate name runs afoul of spam filters.
* Release: 29 Oct 2023

## 1.1.0 
Adding config_file_list() and known_domains().
* Release: 6 Nov 2023

## 1.1.1
Address issues with multiple recipients.
* Release: 21 Nov 2023


tests/__init__.py
# from os import path
#
# TEST_CONFIG_DIR = path.join(path.split(__file__)[0], 'test_configs')


tests/test_mail_sender.py
from os import path
# from unittest import TestCase
import unittest
from email.mime.base import MIMEBase
import platformdirs

import configured_mail_sender
from configured_mail_sender.mail_sender import MailSender,\
    create_sender, MailSenderException
from unittest import TestCase

TEST_CONFIG_DIR = path.join(path.split(__file__)[0], 'test_configs')

CONF_BASE = TEST_CONFIG_DIR
CONF_FILE = path.join(CONF_BASE, 'mailsender_domains.yml')
# APPLICATION = 'testapp'
SENDER = 'foo@base.com'
APPLICATION_PARAM = "test_setting"
APPLICATION_SETTING_VALUE = "setting_value"
SERVICE_NAME = "test_service"


class TestBase(MailSender):
    def __init__(self, sender, **kwargs):
        MailSender.__init__(self, sender, **kwargs)
        self.message = None

    def get_service_name(self):
        return SERVICE_NAME

    def send_message(self, message: MIMEBase) -> None:
        self.message = message


class Test(TestCase):
    def test_creation(self):
        """Instantiate basic class and verify some generic settings"""
        mailer = create_sender(SENDER, overrides=CONF_FILE)

        self.assertEqual(mailer.sender, SENDER)
        self.assertEqual(SERVICE_NAME, mailer.get_service_name())

    def test_no_gmail(self):
        """Make sure attempt to use gmail protocol fails as expected"""
        self.assertRaises(MailSenderException,
                          lambda: create_sender("foo@gmail.none.test",
                                               overrides=CONF_FILE))

    def test_bad_module(self):
        self.assertRaises(MailSenderException,
                          lambda: create_sender('foo@bad.server',
                                                overrides=CONF_FILE))

    def test_bad_domain(self):
        self.assertRaises(MailSenderException,
                          lambda: create_sender("foo@nosuchdomain"))

    def test_no_sender(self):
        self.assertRaises(MailSenderException,
                          lambda: create_sender(None))

    def test_domain_list(self):
        domains = configured_mail_sender.mail_sender.known_domains()
        self.assertEqual('smtp.mail.yahoo.com', domains.get('yahoo.com'))
        self.assertEqual(7, len(domains))

    def test_file_list(self):
        config_files = configured_mail_sender.config_file_list()
        self.assertEqual(path.join(platformdirs.user_config_path('MailSender'),
                                   'mailsender_creds.yml'),
                         config_files[-1])
        self.assertEqual(path.join(platformdirs.site_config_path('MailSender'),
                                   'mailsender_domains.yml'),
                         config_files[0])


if __name__ == '__main__':
    unittest.main(verbosity=2)


tests/test_smtp_sender.py
import unittest
import os
import yaml
from configured_mail_sender.mail_sender import create_sender, MailSenderException
from configured_mail_sender.smtp_sender import (SMTPSender,
                                                SecurityProtocol,
                                                _get_port_security)
from unittest import TestCase
from smtplib import SMTP_SSL
from email.mime.multipart import MIMEMultipart
from email.mime.text import MIMEText

CONF_BASE = os.path.join(os.path.split(__file__)[0], 'test_configs')
SERVICE_NAME = "smtp_test"
SERVICE_DOMAIN = "smtp.test"
SENDER = f"smtp_sender@{SERVICE_DOMAIN}"
SENDER_PASSWORD = "test?pwd9"
PREVIOUS_SENDER = f'prev_sender@{SERVICE_DOMAIN}'
PREVIOUS_PASSWORD = 'test?pwd9'
DOMAIN_FILE = os.path.join(CONF_BASE, "mailsender_domains.yml")


class TestSender(SMTPSender):
    def __init__(self, sender, domain_spec, **kwargs):
        SMTPSender.__init__(self, sender, domain_spec, **kwargs)

    def open(self):
        return SMTPSender.open(self)

    def get_service_name(self) -> str:
        return SERVICE_NAME

    def _create_server(self, **kwargs) -> SMTP_SSL:
        return MockSMTPSSL(**kwargs)

    def _get_password(self):
        # Don't want to have to type a password for tests
        if 'password' in self.kwargs.keys():
            return self.kwargs.get('password')

        return SENDER_PASSWORD


class MockSMTPSSL(SMTP_SSL):
    def __init__(self, **kwargs):
        self.port = kwargs.get('port')
        # self.userid = userid
        self.server = kwargs.get("host")
        self.sender = None
        self.password = None
        self.sent_sender = None
        self.receivers = None
        self.message = None

    def login(self, user, password, *, initial_response_ok=True):
        # verify that user and password are as expected
        self.sender = user
        self.password = password
        return None

    def sendmail(self, from_addr, to_addrs, msg, mail_options=(),
                 rcpt_options=()):
        self.sent_sender = from_addr
        self.receivers = to_addrs
        self.message = msg


class TestInvokeSMTP(TestCase):
    def test_create(self) -> None:
        self.assertRaises(MailSenderException,
                          lambda: create_sender("nobody@yahoo.com",
                                                password="badpwd"))


class TestBasic(TestCase):
    def setUp(self) -> None:
        self.sender = create_sender(SENDER,
                                    overrides=DOMAIN_FILE,
                                    password=SENDER_PASSWORD)
        with open(DOMAIN_FILE, 'r') as f:
            self.service_params = yaml.safe_load(f).get(SERVICE_DOMAIN)

    def test_create(self) -> None:
        """Basic test of mailer creation with password already known"""
        sender = self.sender
        self.assertEqual(SENDER_PASSWORD, sender.password)
        smtp = sender.smtp
        self.assertEqual(SENDER, smtp.sender)
        self.assertEqual(SENDER_PASSWORD, smtp.password)
        self.assertEqual(self.service_params['server'], smtp.server)
        self.assertEqual(self.service_params['port'], smtp.port)

    def test_send(self) -> None:
        """Create a message with To and Bcc and make sure things work as designed."""
        msg = MIMEMultipart()
        msg['Subject'] = 'This is a test'
        test_to = "test_to@example.com"
        test_bcc = "test_bcc@example.com,test_bcc2@example.com"
        test_cc = "test_cc@example.com,test_cc2@example.com"
        msg['To'] = test_to
        msg['Bcc'] = test_bcc
        msg['Cc'] = test_cc

        msg['From'] = SENDER
        msg.attach(MIMEText("Here's your report.", 'plain'))
        self.sender.send_message(msg)

        smtp = self.sender.smtp
        self.assertEqual(smtp.sender, SENDER)
        self.assertEqual(smtp.message, msg.as_string())
        receivers = set(smtp.receivers)
        for rs in [test_to, test_bcc, test_cc]:
            rs = rs.split(',')
            for r in rs:
                self.assertTrue(r in receivers, f'{r} missing in receivers')
                receivers.remove(r)
        self.assertTrue(len(receivers) == 0, f'Extra receivers in {receivers}')


class TestNewPassword(TestCase):

    def setUp(self) -> None:
        # Some setup, and make sure there isn't a leftover password file
        self.other_sender = f'anotherSender@{SERVICE_DOMAIN}'
        self.user_password_file = os.path.join(CONF_BASE,
                                               f'user_{self.other_sender}.yml')
        self.test_cred_file = os.path.join(CONF_BASE, "test_creds.yml")
        if os.path.exists(self.user_password_file):
            os.remove(self.user_password_file)

    def tearDown(self) -> None:
        torm = [f'{f}.lock' for f in [self.test_cred_file,  self.user_password_file]]
        torm += [self.user_password_file]
        # if os.path.exists(self.user_password_file):
        #     os.remove(self.user_password_file)
        for f in torm:
            if os.path.exists(f):
                os.remove(f)

    # If no password is available, a credentials file should be created
    def test_new_password(self):
        """ Verify credentials file update when prompted for a new password. """
        sender = create_sender(self.other_sender,
                               overrides=DOMAIN_FILE,
                               creds_file=self.user_password_file)
        self.assertIsNotNone(sender)
        self.assertTrue(os.path.exists(self.user_password_file),
                        'creds file should be created')
        with open(self.user_password_file, 'r') as f:
            new_certs = yaml.safe_load(f)
        cert_section = new_certs.get(self.other_sender)
        self.assertNotEqual(cert_section, None)
        self.assertEqual(cert_section.get('password'), SENDER_PASSWORD)

    def test_prev_password(self):
        """ Verifying that password in the cred file is correctly retrieved."""
        sender = create_sender(PREVIOUS_SENDER,
                               overrides=DOMAIN_FILE,
                               creds_file=self.test_cred_file)
        self.assertEqual(sender.password, PREVIOUS_PASSWORD)

    def test_explicit_password(self):
        """ Verify that explicit password overrides configuration file """
        test_pwd = "plover"
        sender = create_sender(PREVIOUS_SENDER,
                               overrides=DOMAIN_FILE,
                               password=test_pwd,
                               creds_file=self.test_cred_file)
        self.assertEqual(sender.password, test_pwd)


class TestPortSecurity(TestCase):
    """
    Testing that port or protocol can be set from the other
    """

    def test_bad_security(self):
        """Invalid security parameter should throw an exception"""
        conf = {'security': "unk"}
        self.assertRaises(MailSenderException,
                          lambda: _get_port_security(conf))

    def test_implied_port(self):
        """Port correctly deduced from specified security"""
        for test_security in SecurityProtocol:
            (port, security) = _get_port_security({'security': test_security.name})
            self.assertEqual(test_security.value, port)

    def test_implied_security(self):
        """Security protocol correctly deduced from specified port"""
        for test_security in SecurityProtocol:
            (port, security) = _get_port_security({'port': test_security.value})
            self.assertEqual(test_security.name, security.name)

    def test_both_set(self):
        """If both port and security protocol given, believe the user"""
        (port, security) = _get_port_security({'port': 33,
                                               'security': SecurityProtocol.SSL.name})
        self.assertEqual(33, port)
        self.assertEqual(SecurityProtocol.SSL.name, security.name)

    def test_default_security(self):
        """Assume SSL if using nonstandard port"""
        (port, security) = _get_port_security({'port': 99})
        self.assertEqual(SecurityProtocol.SSL.name, security.name)

    def test_explicit_override(self):
        """Port and security explicitly overwritten by caller"""
        kwargs = {'port': 999, 'security': SecurityProtocol.SSL.name}
        port_sec = {'port': 25, 'security': SecurityProtocol.NONE.name}
        (port, security) = _get_port_security(port_sec, **kwargs)
        self.assertEqual(port, 999)
        self.assertEqual(SecurityProtocol.SSL.name, security.name)


if __name__ == '__main__':
    unittest.main(verbosity=2)




configured_mail_sender/smtp_sender.py
from typing import Tuple

from configured_mail_sender.mail_sender import MailSender, MailSenderException
from getpass import getpass
from smtplib import SMTP_SSL, SMTPException, SMTP
from email.mime.base import MIMEBase
from enum import Enum

"""
This is a MailSender version for sending to a generic SMTP server. It assumes
SSL is needed.

This needs:
    server, port:
        The server and port to connect to. These could be passed in as parameters,
        or could come from the caller, or could be taken from the domain configuration
        file. We assume that server and port will be the same for any application.
    
    sender:
        The email sender, who's sending the email. (User and domain.) This is an
        explicit parameter from the mail_sender() caller.
    
    password:
        This comes from the user's readonly credentials file file. If the file
        doesn't exist with the password, SMTPSender prompts the user to enter a
        password and saves the password for later use.
    
SMTP supports three encryption protocols, and by convention those correspond
to standard ports:
    port 25 implies no encryption. Connections to that port are vulnerable to
        sniffing data, including your password. Its use is generally not supported
        and is discouraged, but there may be some servers that still use it.
    Port 465 is SSL encrypted, so much more secure, but there are vulnerabilities.
        But this is the most common implementation.
    Port 587 is encrypted but uses STARTTLS, which avoids the SSL vulnerability.
        But it isn't always supported.
    
When creating a connection, if only port or security is specified in the domain
configuration, the other parameter will be inferred from the first. If neither
is given, SSL on port 465 is used.
"""


class SecurityProtocol(Enum):
    NONE = 25  # Is this really supported
    SSL = 465
    STARTTLS = 587


# Map from standard port to default protocol.
_PORT_SECURITY = {p.value: p for p in SecurityProtocol}


def _get_port_security(domain_spec: dict,
                       **kwargs: dict) -> Tuple[int, SecurityProtocol]:
    """Return the port security enum from the configuration
    :param domain_spec: Configuration for this email domain
    :param **port_spec: Override port in domain_spec
    :param **security: Override security in domain_spec
    :return: Port number and security
    """

    port = kwargs.get('port', domain_spec.get('port'))
    security_name = kwargs.get('security', domain_spec.get('security'))
    if security_name:
        try:
            security = SecurityProtocol[security_name.upper()]
            if not port:
                # Port not specified, so use default for the protocol
                port = security.value
        except KeyError as e:
            raise MailSenderException(e,
                                      f'{security_name} is not one of '
                                      f'{",".join(p.name for p in SecurityProtocol)}')
    else:
        port = port if port else SecurityProtocol.SSL.value
        security = _PORT_SECURITY.get(port, SecurityProtocol.SSL)

    return port, security


class SMTPSender(MailSender):
    def __init__(self, sender: str,
                 domain_spec: dict,
                 **kwargs: dict):
        """
        Class to facilitate sending gmail
        :param sender: Sending email address
        :param **server: Server name
        :param **password: Password for connection if explicitly provided
        """
        MailSender.__init__(self, sender, domain_spec=domain_spec, **kwargs)

        server = self.domain_spec.get('server')
        (self.port, self.security) = _get_port_security(self.domain_spec, **kwargs)

        # Not recommended, but if password came in from creation, save it
        self.password = kwargs.get('password')
        if self.password:
            self.user_credentials['password'] = self.password

        self.smtp = None
        self.server = server
        self.userid = None

    def open(self):
        """Initialize, including setting up connection to the server.
        :returns: self
        """
        MailSender.open(self)
        # Now that we know who to connect to, get the user's password.
        update_needed = self._complete_credentials()
        self.smtp = self._create_server(host=self.server, timeout=50, port=self.port)

        self.userid = self.kwargs.get('userid',
                                      self.user_credentials.get('userid', self.sender))
        try:
            self.smtp.login(self.userid, self.password)
        except SMTPException as e:
            raise MailSenderException(e, f'Login to {self.userid} failed.')

        # Successfully logged in. Update credential file if needed
        if update_needed:
            self._update_creds()
        return self

    def _create_server(self, **kwargs) -> SMTP:
        """Open an SMTP connection. (This is a method to facilitate unit tests.)
        :param kwargs: Other parameters including host and port.
        :return: SMTP Server
        """
        if self.security is SecurityProtocol.SSL:
            return SMTP_SSL(**kwargs)
        else:
            sender = SMTP(**kwargs)
            if self.security is SecurityProtocol.STARTTLS:
                sender.starttls()
            return sender

    def _complete_credentials(self) -> bool:
        """
        Initialize credentials from credentials file.
        Extending class may need to add more than password?
        :return: True if credentials file needs to be re-written.
        """

        if self.password:
            # We already have a password
            return False

        self.password = self.user_credentials.get('password', None)
        if self.password:
            return False
        else:
            # Need to get password and save it if valid.
            self.password = self.user_credentials['password'] = self._get_password()
            return True

    def _get_password(self) -> str:
        """
        Ask the user to enter a password.
        :return: password(str)

        Note: This is a non-static method to facilitate unit tests.
        """

        # Not recommended, mostly for unit tests, but use password from
        # instantiation if specified.
        pwd = self.kwargs.get("password")
        if pwd is None:
            pwd = getpass(f'Enter password for user "{self.sender}:')
        return pwd

    def send_message(self, message: MIMEBase) -> None:
        """
        Send a MIME email message. Recipients are encoded in the message.
        :param message: message to send
        """
        r_list = [message.get(r) for r in ['To', 'Cc', 'Bcc']]
        r_list = [r for r in r_list if r]
        receivers = ','.join(r_list)
        message['From'] = self.sender
        self._send_message(self.sender, receivers, message.as_string())

    def _send_message(self, sender: str, receivers: str, msg_str: str) -> None:
        """
        Do the actual send. (This is a class method to facilitate unit tests.)
        :param sender: Sending address
        :param receivers: Comma-separated list of receivers
        :param msg_str: Message as encoded string
        """
        self.smtp.sendmail(sender, receivers.split(','), msg_str)

    def get_service_name(self) -> str:
        return 'smtp'


configured_mail_sender/mail_sender.py
import os
import platformdirs
import yaml
from abc import abstractmethod
from email.mime.base import MIMEBase
from os import environ, path
from typing import Union, MutableMapping, List, Dict, Mapping
import combine_settings
from filelock import FileLock

_OUTLOOK_SERVER = {
    'server': 'smtp-mail.outlook.com',
    'port': 587,
}
_BUILTIN_DOMAINS = {
    'yahoo.com': {
        'protocol': 'smtp',
        'server': 'smtp.mail.yahoo.com',
        'port': 587,
    },
    'aol.com': {
        'server': 'smtp.aol.com',
        'port': 465,
    },
    'gmail.com': {
        'server': 'smtp.gmail.com',
    },
    'outlook.com': _OUTLOOK_SERVER,
    'hotmail.com': _OUTLOOK_SERVER,
    'live.com': _OUTLOOK_SERVER,
    'comcast.net': {
        'server': 'smtp.comcast.net',
        'port': 587,
    },
}

CONFIG_FILE_NAME = 'mailsender_domains.yml'
CREDS_FILE_NAME = 'mailsender_creds.yml'
CONFIG_APPLICATION_NAME = 'MailSender'
DEFAULT_CREDS_FILE = path.join(platformdirs.user_config_dir(CONFIG_APPLICATION_NAME),
                               CREDS_FILE_NAME)


class MailSenderException(Exception):
    def __init__(self, *args, **kwargs):
        Exception.__init__(self, *args, **kwargs)


class MailSenderUnsupportedException(MailSenderException):
    def __init__(self, *args, **kwargs):
        MailSenderException.__init__(self, *args, **kwargs)


class MailSender:
    def __init__(self,
                 sender: str,
                 domain_spec: Dict,
                 **kwargs: Dict):
        """
        Generic (abstract) MailSender - Class to send email from a specified sender
        :param sender str: Email address we'll be sending as
        :param domain_spec dict: Settings for this domain
        :param **creds_file str: Path to user credentials file
        """

        # Save these in case someone wants other parameters
        self.kwargs = kwargs

        # This has domain-specific settings
        self.domain_spec = domain_spec

        # Load user service-specific credentials (passwords or whatever).
        # User credentials are from the first specified of:
        #   - The creds_file parameter.
        #   - The MAILSENDER_CREDS environment variable
        #   - mailsender_creds.yml in the user's standard configuration file
        # default_file =
        #             path.join(platformdirs.user_config_path(CONFIG_APPLICATION_NAME),
        #                          'mailsender_creds.yml')
        self.user_cred_file = _get_user_cred_file(**kwargs)
        creds_dir = path.split(self.user_cred_file)[0]
        # Make sure the credentials directory exists, even with no actual credentials.
        os.makedirs(creds_dir, mode=0o700, exist_ok=True)

        self.user_cred_lock = f'{self.user_cred_file}.lock'
        self.cred_locker = FileLock(self.user_cred_lock)
        self.user_credentials = self._read_creds_file().get(sender)
        if not self.user_credentials:
            self.user_credentials = {}
        self.sender = sender

    def open(self) -> 'MailSender':
        """Do initialization. Variables can be initialized in __init__, but any
        interactions with externals should happen here.

        :returns: self
        """
        return self

    def _read_creds_file(self) -> MutableMapping:
        """
        Return current content of creds file.
        :return: Content of creds file
        """
        with self.cred_locker:
            if path.exists(self.user_cred_file):
                try:
                    with open(self.user_cred_file, 'r') as f:
                        # Question: is there a way to tell safe_load to allow tabs?
                        creds = yaml.load(f, Loader=yaml.Loader)
                    return creds
                except IOError as e:
                    raise MailSenderException(e, f"Error opening {self.user_cred_file}")
            else:
                return {}

    def _update_creds(self) -> None:
        """
        Update the user credentials file from the current content of user_creds.
        :return: None
        """
        with self.cred_locker:
            # First, reload old credentials
            current_creds = self._read_creds_file()
            current_creds[self.sender] = self.user_credentials

            temp_file = f'{self.user_cred_file}.temp'
            try:
                # Write the new file, then replace the old file, to avoid damaging
                # the old file if something breaks in the middle.
                fd = os.open(temp_file, (os.O_CREAT | os.O_TRUNC | os.O_WRONLY), 0o600)
            except IOError as e:
                raise MailSenderException(e, f"Can't write {self.user_cred_file}")

            try:
                os.write(fd, bytes(yaml.safe_dump(current_creds), 'utf-8'))
                os.replace(temp_file, self.user_cred_file)
            except Exception as e:
                raise MailSenderException(e,
                                          "Problem writing new credentials" +
                                          temp_file)
            finally:
                os.close(fd)

    @abstractmethod
    def send_message(self, message: MIMEBase) -> None:
        """
        Send an email message. (Recipients are encoded in the message.)
        :param message: Message to send
        :return: None
        """

    @abstractmethod
    def get_service_name(self) -> str:
        """Return name of service.
        Extending class should override this to return
        the appropriate service name.

        :return: Name of service
        """


# TODO: Merge smtp_sender into here?


def create_sender(sender: str,
                  base_config: Union[Dict, str] = None,
                  overrides: Union[Dict, str] = None,
                  **kwargs) -> MailSender:
    """
    Create a MailSender instance to send email from the given sender from a
    given user.
    :param sender: Sending email address
    :param base_config: Override built-in base configuration for load_config()
    :param overrides: Settings to override default load in load_config()
    :param kwargs: Other settings that might be needed for specific MailSenders
    :return: A MailSender instance
    :raises: MailSenderException
    """

    if not sender:
        raise MailSenderException('sender is required')

    # Get server configurations (service type, SMTP url and port, etc)
    domains_conf = _get_domain_conf(base_config=base_config,
                                   overrides=overrides)

    # # Load global configurations (service type, SMTP url and port, etc)
    # domains_conf = combine_settings.load_config(CONFIG_FILE_NAME,
    #                                             base_config=base_config,
    #                                             application=CONFIG_APPLICATION_NAME,
    #                                             overrides=overrides)

    (addr, domain) = sender.split('@', 2)
    domain_spec = domains_conf.get(domain)
    if not domain_spec:
        raise MailSenderException(f"Domain {domain} isn't recognized")

    # Primarily for testing, you can provide an explicit "protocol" class specifying
    # the MailSender implementation to instantiate.
    protocol = domain_spec.get('protocol', 'smtp')

    # Will add this back if/when I add gmail OAuth2 support
    # if protocol == 'gmail':
    #   try:
    #     from gmail_sender import gmail_sender
    #     return gmail_sender.GmailSender(sender,
    #                                     domain_spec=domain_spec, **kwargs).open()
    #   except ModuleNotFoundError as e:
    #     raise MailSenderUnsupportedException(e,
    #                                        "mail_sender_gmail package not installed")

    if protocol == 'smtp':
        from configured_mail_sender import smtp_sender
        return smtp_sender.SMTPSender(sender, domain_spec=domain_spec, **kwargs).open()

    if ':' not in protocol:
        raise MailSenderException(f'No implementation specified for domain {domain}')

    # Maybe an explicit module:class?
    (module_, class_) = protocol.split(':', 1)
    try:
        import importlib
        mod = importlib.import_module(module_)
        cls = getattr(mod, class_)
        inst = cls(sender, domain_spec=domain_spec, **kwargs)
        if not isinstance(inst, MailSender):
            raise MailSenderException(f'{module_}:{class_} for '
                                      f'{domain} is not a MailSender')
        return inst.open()
    except (ModuleNotFoundError, AttributeError) as e:
        raise MailSenderUnsupportedException(e,
                                             f'{module_}:{class_} '
                                             f'for email domain {domain} unknown)')


def _get_user_cred_file(creds_file=None,
                        **kwargs) -> str:
    """
    Return path to user credentials file
    :param creds_file: User-specified file
    :return: Creds file
    """
    if not creds_file:
        creds_file = environ.get('MAILSENDER_CREDS', DEFAULT_CREDS_FILE)
    return creds_file


def config_file_list(base_config: str = None,
                     overrides: str = None,
                     creds_file: str = None) -> List[str]:
    """
    Get configuration file list
    :param base_config: Optional base configuration for load_config
    :param overrides: An additional file with domain specifications
    :param creds_file: Explicit path to user credentials file
    :return: List of directories searched
    """
    file_list = combine_settings.config_file_list(
        CONFIG_FILE_NAME,
        base_config=base_config,
        application=CONFIG_APPLICATION_NAME,
        overrides=overrides)

    file_list.append(_get_user_cred_file(creds_file))
    return file_list


def _get_domain_conf(base_config: Union[str, Dict] = None,
                     overrides: Union[str, Dict] = None) -> Mapping[str, Dict]:

    base_config = base_config if base_config else _BUILTIN_DOMAINS
    # Load global configurations (service type, SMTP url and port, etc)
    return combine_settings.load_config(CONFIG_FILE_NAME,
                                        base_config=base_config,
                                        application=CONFIG_APPLICATION_NAME,
                                        overrides=overrides)


def known_domains(base_config: str = None,
                  overrides: str = None) -> Dict[str, str]:
    """
    Return domain to server mapping for known domains
    :param base_config: Alternate base configuration
    :param overrides: User-specified override file
    :return: Mapping from domain to server
    """
    domains_conf = _get_domain_conf(base_config=base_config,
                                    overrides=overrides)
    ret = {}
    for domain, params in domains_conf.items():
        protocol = params.get('protocol', 'smtp')
        ret[domain] = params.get('server') if protocol == 'smtp' else protocol
    return ret


if __name__ == '__main__':
    pass
